coffee_core/src/coffee.rs

```diff
@@ -272,11 +273,77 @@ impl PluginManager for CoffeeManager {
         })
     }
 
-    async fn upgrade(&mut self, _: &[&str]) -> Result<(), CoffeeError> {
+    async fn upgrade_single(&mut self, repo: &str) -> Result<CoffeeUpgradeOne, CoffeeError> {
+        let status;
+        match self.repos.get(repo) {
+            Some(repo) => {
+                // get the list of the plugins installed from this repository
+                // TODO: add a field of installed plugins in the repository struct instead
+                let mut installed_plugins: Vec<String> = vec![];
+                let remote_repo = repo.list().await?;
+                let plugins = self.config.plugins.clone();
+                for plugin in &remote_repo {
+                    if let Some(ind) = plugins
+                        .iter()
+                        .position(|elem| elem.name() == *plugin.name())
+                    {
+                        let plugin_name = &plugins[ind].name().clone();
+                        installed_plugins.push(plugin_name.to_owned());
+                    }
+                }
+                // pull the changes from the repository
+                status = repo.pull().await?;
+                match status {
+                    UpgradeStatus::UpToDate => {
+                        log::debug!("repository up to date: {}", repo.name());
+                    }
+                    UpgradeStatus::Updated => {
+                        // uninstall the plugins
+                        // for plugin in &installed_plugins {
+                        //     self.remove(&plugin).await?;
+                        // }
+                        // reinstall the plugins
+                        // for plugin in &installed_plugins {
+                        //     self.install(&plugin, false, false).await?;
+                        // }
+                        log::debug!("repository upgraded: {}", repo.name());
+                    }
+                }
+                self.storage.store(&self.storage_info()).await?;
+            }
+            None => {
+                return Err(error!("repository with name: {repo} not found"));
+            }
+        };
+
+        Ok(CoffeeUpgradeOne {
+            repo: repo.to_owned(),
+            status,
+        })
+    }
+
+    async fn upgrade(&mut self, repo: &str, all: bool) -> Result<CoffeeUpgrade, CoffeeError> {
         self.remote_sync().await?;
-        // FIXME: Fix debug message with the list of plugins to be upgraded
-        log::debug!("upgrading plugins");
-        Ok(())
+        let mut total_status: Vec<CoffeeUpgradeOne> = vec![];
+        if all {
+            // for repository in self.repos.keys() {
+            //     match self.upgrade_single(repository).await {
+            //         Ok(val) => total_status.push(val),
```

Comment:
Feedback that I had from a design point of view, it is possible to write the `async fn upgrade_single(&mut self, repo: &str) -> Result<CoffeeUpgradeOne, CoffeeError> {` inside the single repository?

This should remove a borrowing problem, but also this is the right design because the repository knows how to upgrade itself.

The only think that I am not sure is we the repository to upgrade itself need to know somethings from outside the repository.

Sorry I did not look at the code yet